### PR TITLE
[IOTDB-839] Make Tablet api more friendly

### PR DESCRIPTION
**Pull request by wyh-ztf**

### CODE REVIEW

File: tsfile/src/main/java/org/apache/iotdb/tsfile/write/record/Tablet.java [hunk 1]
```diff
@@ -81,6 +82,43 @@ public Tablet(String deviceId, List<MeasurementSchema> timeseries) {
     this(deviceId, timeseries, DEFAULT_SIZE);
   }
 
+  public void addTimestamp(int rowIndex, long timestamp) {
+    timestamps[rowIndex - 1] = timestamp;
+  }
+
+  public void addValue(MeasurementSchema measurementSchema, int rowIndex, Object value) {
+
+    int indexOfValue = schemas.indexOf(measurementSchema);
```
Comment (mychaow) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/write/record/Tablet.java:91]: It's not a good way to get the index, measurements maybe very large. You should store the measurementSchema as a map

File: tsfile/src/main/java/org/apache/iotdb/tsfile/write/record/Tablet.java [hunk 2]
```diff
@@ -81,6 +82,43 @@ public Tablet(String deviceId, List<MeasurementSchema> timeseries) {
...
+    if (measurementSchema.getType() == INT64) {
+      long[] sensor = (long[]) values[indexOfValue];
+      sensor[rowIndex - 1] = (long) value;
+      values[indexOfValue] = sensor;
```
Comment (mychaow) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/write/record/Tablet.java:95]: what is the meaning of this line?

File: tsfile/src/main/java/org/apache/iotdb/tsfile/write/record/Tablet.java [hunk 3]
```diff
@@ -81,6 +82,43 @@ public Tablet(String deviceId, List<MeasurementSchema> timeseries) {
     this(deviceId, timeseries, DEFAULT_SIZE);
   }
 
+  public void addTimestamp(int rowIndex, long timestamp) {
```
Comment (mychaow) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/write/record/Tablet.java:85]: rowIndex maybe starts with 0?

File: tsfile/src/main/java/org/apache/iotdb/tsfile/write/record/Tablet.java [hunk 4]
```diff
@@ -95,12 +103,48 @@ public Tablet(String deviceId, List<MeasurementSchema> schemas, int maxRowNumber
...
+    timestamps[rowIndex] = timestamp;
+  }
+
+  public void addValue(MeasurementSchema measurementSchema, int rowIndex, Object value) {
+    int indexOfValue =mschemawithindex.get(measurementSchema);
+
+    if (measurementSchema.getType() == INT64) {
```
Comment (qiaojialin) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/write/record/Tablet.java:124]: switch (measurementSchema.getType())

File: tsfile/src/main/java/org/apache/iotdb/tsfile/write/record/Tablet.java [hunk 5]
```diff
@@ -95,12 +103,48 @@ public Tablet(String deviceId, List<MeasurementSchema> schemas, int maxRowNumber
     this.deviceId = deviceId;
     this.schemas = schemas;
     this.maxRowNumber = maxRowNumber;
+    mschemawithindex=new HashMap<>();
+
+    for(int i=0;i<schemas.size();i++){
```
Comment (qiaojialin) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/write/record/Tablet.java:108]: There is a java-google-style.xml under the root folder, import it as the default code style and format the code.

It should be something like this:
for(int i = 0; i < schemas.size(); i++){

File: tsfile/src/main/java/org/apache/iotdb/tsfile/write/record/Tablet.java [hunk 6]
```diff
@@ -95,12 +103,48 @@ public Tablet(String deviceId, List<MeasurementSchema> schemas, int maxRowNumber
...
     createColumns();
 
     reset();
   }
 
+  public void addTimestamp(int rowIndex, long timestamp) {
+    timestamps[rowIndex] = timestamp;
+  }
+
+  public void addValue(MeasurementSchema measurementSchema, int rowIndex, Object value) {
```
Comment (qiaojialin) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/write/record/Tablet.java:121]:
```suggestion
  public void addValue(String measurement, int rowIndex, Object value) {
```

Using measurement id may be more friendly to users.

**Review by qiaojialin — COMMENTED** [review]
Good work! It's better to give two examples of constructing a tablet in SessionExample to let users know the old interface is still available.

File: example/session/src/main/java/org/apache/iotdb/SessionExample.java [hunk 7]
```diff
@@ -247,15 +247,11 @@ private static void insertTablet() throws IoTDBConnectionException, StatementExe
 
     Tablet tablet = new Tablet("root.sg1.d1", schemaList, 100);
 
-    long[] timestamps = tablet.timestamps;
```
Comment (mychaow) [inline example/session/src/main/java/org/apache/iotdb/SessionExample.java:250]: You should not delete the previous example, user also can use the previous way to add data.  You should add new code to show how to use.

File: example/session/src/main/java/org/apache/iotdb/SessionExample.java [hunk 8]
```diff
@@ -247,15 +247,11 @@ private static void insertTablet() throws IoTDBConnectionException, StatementExe
...
     for (long time = 0; time < 100; time++) {
-      int row = tablet.rowSize++;
-      timestamps[row] = time;
-      for (int i = 0; i < 3; i++) {
-        long[] sensor = (long[]) values[i];
-        sensor[row] = i;
+      int rowIndex = tablet.rowSize++;
+      tablet.addTimestamp(rowIndex, time);
+      for (int s = 0; s < 3; s++) {
+        tablet.addValue(schemaList.get(s), rowIndex, (long) s);
```
Comment (mychaow) [inline example/session/src/main/java/org/apache/iotdb/SessionExample.java:254]: It's not nesscessary to cast to long. You have done cast in the addValue.

File: session/src/test/java/org/apache/iotdb/session/IoTDBSessionIT.java [hunk 9]
```diff
@@ -812,15 +811,11 @@ private void insertTabletTest1(String deviceId)
...
     for (long time = 0; time < 100; time++) {
-      int row = tablet.rowSize++;
-      timestamps[row] = time;
-      for (int i = 0; i < 3; i++) {
-        long[] sensor = (long[]) values[i];
-        sensor[row] = i;
+      int rowIndex = tablet.rowSize++;
+      tablet.addTimestamp(rowIndex, time);
+      for (int s = 0; s < 3; s++) {
+        tablet.addValue(schemaList.get(s), rowIndex, (long) s);
```
Comment (mychaow) [inline session/src/test/java/org/apache/iotdb/session/IoTDBSessionIT.java:818]: same with the previous comment

File: tsfile/src/main/java/org/apache/iotdb/tsfile/write/record/Tablet.java (hunk 5 above)
Comment (mychaow) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/write/record/Tablet.java:108]: code should be 
`for (int i = 0; i < schemas.size(); i++) {}`.  It's not wasteful to add blank in the code.

File: tsfile/src/main/java/org/apache/iotdb/tsfile/write/record/Tablet.java [hunk 10]
```diff
@@ -95,12 +103,48 @@ public Tablet(String deviceId, List<MeasurementSchema> schemas, int maxRowNumber
...
+    int indexOfValue =mschemawithindex.get(measurementSchema);
```
Comment (mychaow) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/write/record/Tablet.java:122]: add blank

File: example/session/src/main/java/org/apache/iotdb/SessionExample.java [hunk 11]
```diff
@@ -247,15 +247,11 @@ private static void insertTablet() throws IoTDBConnectionException, StatementExe
...
-
     for (long time = 0; time < 100; time++) {
-      int row = tablet.rowSize++;
-      timestamps[row] = time;
-      for (int i = 0; i < 3; i++) {
-        long[] sensor = (long[]) values[i];
-        sensor[row] = i;
+      int rowIndex = tablet.rowSize++;
+      tablet.addTimestamp(rowIndex, time);
+      for (int s = 0; s < 3; s++) {
```
Comment (neuyilan) [inline example/session/src/main/java/org/apache/iotdb/SessionExample.java:253]: I think it's better to separate the rowIndex, the time column, and the value column, which would clearer to users.

File: tsfile/src/main/java/org/apache/iotdb/tsfile/write/record/Tablet.java (hunk 4 above)
Comment (wyh-ztf) [reply tsfile/src/main/java/org/apache/iotdb/tsfile/write/record/Tablet.java:124]: Thank you. I'll improve it

File: tsfile/src/main/java/org/apache/iotdb/tsfile/write/record/Tablet.java (hunk 5 above)
Comment (wyh-ztf) [reply tsfile/src/main/java/org/apache/iotdb/tsfile/write/record/Tablet.java:108]: copy that,thank you

File: tsfile/src/main/java/org/apache/iotdb/tsfile/write/record/Tablet.java (hunk 6 above)
Comment (wyh-ztf) [reply tsfile/src/main/java/org/apache/iotdb/tsfile/write/record/Tablet.java:121]: copy that,thank you

File: example/session/src/main/java/org/apache/iotdb/SessionExample.java (hunk 8 above)
Comment (wyh-ztf) [reply example/session/src/main/java/org/apache/iotdb/SessionExample.java:254]: copy that,thank you

File: session/src/test/java/org/apache/iotdb/session/IoTDBSessionIT.java (hunk 9 above)
Comment (wyh-ztf) [reply session/src/test/java/org/apache/iotdb/session/IoTDBSessionIT.java:818]: copy that,thank you

File: tsfile/src/main/java/org/apache/iotdb/tsfile/write/record/Tablet.java (hunk 5 above)
Comment (wyh-ztf) [reply tsfile/src/main/java/org/apache/iotdb/tsfile/write/record/Tablet.java:108]: copy that,thank you

File: tsfile/src/main/java/org/apache/iotdb/tsfile/write/record/Tablet.java (hunk 10 above)
Comment (wyh-ztf) [reply tsfile/src/main/java/org/apache/iotdb/tsfile/write/record/Tablet.java:122]: copy that,thank you

File: example/session/src/main/java/org/apache/iotdb/SessionExample.java (hunk 11 above)
Comment (wyh-ztf) [reply example/session/src/main/java/org/apache/iotdb/SessionExample.java:253]: copy that,thank you

File: example/session/src/main/java/org/apache/iotdb/SessionExample.java (hunk 7 above)
Comment (wyh-ztf) [reply example/session/src/main/java/org/apache/iotdb/SessionExample.java:250]: copy that,thank you

File: session/src/test/java/org/apache/iotdb/session/IoTDBSessionIT.java [hunk 12]
```diff
@@ -1087,15 +1086,13 @@ private void insertTabletTest2(String deviceId)
...
     for (long time = 0; time < 1000; time++) {
-      int row = tablet.rowSize++;
-      timestamps[row] = time;
-      for (int i = 0; i < 3; i++) {
-        long[] sensor = (long[]) values[i];
-        sensor[row] = i;
+      int rowIndex = tablet.rowSize++;
+      tablet.addTimestamp(rowIndex, time);
+      for (int s = 0; s < 3; s++) {
+        long value = 0;
```
Comment (mychaow) [inline session/src/test/java/org/apache/iotdb/session/IoTDBSessionIT.java:1093]: value always is initialize to be 0

File: session/src/test/java/org/apache/iotdb/session/IoTDBSessionIT.java [hunk 13]
```diff
@@ -1119,15 +1116,13 @@ private void insertTabletTest3(String deviceId)
...
     for (long time = 500; time < 1500; time++) {
-      int row = tablet.rowSize++;
-      timestamps[row] = time;
-      for (int i = 0; i < 3; i++) {
-        long[] sensor = (long[]) values[i];
-        sensor[row] = i;
+      int rowIndex = tablet.rowSize++;
+      tablet.addTimestamp(rowIndex, time);
+      for (int s = 0; s < 3; s++) {
+        long value = 0;
```
Comment (mychaow) [inline session/src/test/java/org/apache/iotdb/session/IoTDBSessionIT.java:1123]: same with previous

File: tsfile/src/main/java/org/apache/iotdb/tsfile/write/record/Tablet.java [hunk 14]
```diff
@@ -95,12 +102,56 @@ public Tablet(String deviceId, List<MeasurementSchema> schemas, int maxRowNumber
...
+    timestamps[rowIndex] = timestamp;
+  }
+
+  public void addValue(String measurementId, int rowIndex, Object value) {
+    int indexOfValue = mschemawithindex.get(measurementId);
+    MeasurementSchema measurementSchema = schemas.get(indexOfValue);
+
+    switch (measurementSchema.getType()) {
+      case TEXT:
+        Binary[] sensort = (Binary[]) values[indexOfValue];
```
Comment (mychaow) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/write/record/Tablet.java:126]: It's too ugly to create so many new variable names. 
`you could use {} in every case`

Comment (wyh-ztf) [reply tsfile/src/main/java/org/apache/iotdb/tsfile/write/record/Tablet.java:126]: ok, i will improve that

File: session/src/test/java/org/apache/iotdb/session/IoTDBSessionIT.java (hunk 12 above)
Comment (wyh-ztf) [reply session/src/test/java/org/apache/iotdb/session/IoTDBSessionIT.java:1093]: I added it after the loop

File: tsfile/src/main/java/org/apache/iotdb/tsfile/write/record/Tablet.java [hunk 15]
```diff
@@ -52,6 +54,11 @@
    */
   private List<MeasurementSchema> schemas;
 
+  /**
+   * measurementId->indexOf(measurementSchema)
+   */
+  private Map<String, Integer> mschemawithindex;
```
Comment (mychaow) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/write/record/Tablet.java:60]: mschemawithindex should be camel-case name, like mSchemaWithIndex.

Comment (wyh-ztf) [reply tsfile/src/main/java/org/apache/iotdb/tsfile/write/record/Tablet.java:60]: copy that.

File: session/src/test/java/org/apache/iotdb/session/IoTDBSessionIT.java [hunk 16]
```diff
@@ -622,6 +622,17 @@ public void TestSessionInterfacesWithDisabledWAL()
...
+      int rowIndex = time - 1;
+      tablet.addTimestamp(rowIndex, time);
+      long value = 0;
+      for (int s = 0; s < 3; s++) {
+        tablet.addValue(schemaList.get(s).getMeasurementId(), rowIndex, value);
+        value++;
+      }
+      tablet.rowSize++;
+    }
+
```
Comment (qiaojialin) [inline session/src/test/java/org/apache/iotdb/session/IoTDBSessionIT.java:635]: this test lacks an assert. You can do a select count(*) query and assert

File: session/src/test/java/org/apache/iotdb/session/IoTDBSessionIT.java [hunk 17]
```diff
@@ -1203,7 +1289,7 @@ private void queryForBatch() throws ClassNotFoundException, SQLException {
       }
       Assert.assertEquals(standard, resultStr.toString());
       // d1 and d2 will align
-      Assert.assertEquals(7000, count);
+      Assert.assertEquals(14000, count);
```
Comment (qiaojialin) [inline session/src/test/java/org/apache/iotdb/session/IoTDBSessionIT.java:1292]: It's not intuitive where 14000 comes from.

I feel that the IoTDBSessionIT is out of control gradually... 
It's time to remove duplicated methods such as insertTabletTest1-2-3.  
Besides, there are some commented lines in test(), it should be cleaned.

I create a jira issue:  https://issues.apache.org/jira/browse/IOTDB-856. 
Better to fix it in this PR.

File: tsfile/src/main/java/org/apache/iotdb/tsfile/write/record/Tablet.java [hunk 18]
```diff
@@ -52,6 +54,11 @@
    */
   private List<MeasurementSchema> schemas;
 
+  /**
+   * measurementId->indexOf(measurementSchema)
+   */
+  private Map<String, Integer> mSchemaWithIndex;
```
[the remark this answers is not in the document]
Comment (wyh-ztf) [reply tsfile/src/main/java/org/apache/iotdb/tsfile/write/record/Tablet.java:60]: thanks,i will improve that

File: session/src/test/java/org/apache/iotdb/session/IoTDBSessionIT.java (hunk 16 above)
Comment (wyh-ztf) [reply session/src/test/java/org/apache/iotdb/session/IoTDBSessionIT.java:635]: i will fix it in PR IOTDB-856:Clear up the IoTDBSessionIT

File: session/src/test/java/org/apache/iotdb/session/IoTDBSessionIT.java (hunk 17 above)
Comment (wyh-ztf) [reply session/src/test/java/org/apache/iotdb/session/IoTDBSessionIT.java:1292]: ok,i will fix it in that PR

**Review by qiaojialin — APPROVED** [review]

**Review by mychaow — APPROVED** [review]

**Review by neuyilan — APPROVED** [review]

**Pull request merged**